tortoise/state.go

```diff
@@ -5,78 +5,242 @@ import (
+		historicallyVerified types.LayerID
+		// last processed layer
+		minprocessed, processed types.LayerID
+		// last evicted layer
+		evicted types.LayerID
+
+		localThreshold  weight
+		globalThreshold weight
+
+		// epochWeight average weight per layer of atx's that target keyed epoch
+		epochWeight map[types.EpochID]weight
+		// referenceHeight is a median height from all atxs that target keyed epoch
+		referenceHeight map[types.EpochID]uint64
+		// referenceWeight stores atx weight divided by the total number of eligibilities.
+		// it is computed together with refBallot weight. it is not equal to refBallot
+		// only if refBallot has more than 1 eligibility proof.
+		referenceWeight map[types.BallotID]weight
+
+		layers  map[types.LayerID]*layerInfo
+		ballots map[types.LayerID][]*ballotInfo
+
+		// to efficiently find base and reference ballots
+		ballotRefs map[types.BallotID]*ballotInfo
+		// to efficiently decode exceptions
+		blockRefs map[types.BlockID]*blockInfo
+	}
+)
+
+func newState() *state {
+	return &state{
+		epochWeight:     map[types.EpochID]util.Weight{},
+		referenceHeight: map[types.EpochID]uint64{},
+		referenceWeight: map[types.BallotID]util.Weight{},
+
+		layers:     map[types.LayerID]*layerInfo{},
+		ballots:    map[types.LayerID][]*ballotInfo{},
+		ballotRefs: map[types.BallotID]*ballotInfo{},
+		blockRefs:  map[types.BlockID]*blockInfo{},
+	}
+}
+
+func (s *state) layer(lid types.LayerID) *layerInfo {
+	layer, exist := s.layers[lid]
+	if !exist {
+		layer = &layerInfo{lid: lid, empty: util.WeightFromUint64(0)}
+		s.layers[lid] = layer
+	}
+	return layer
+}
+
+func (s *state) addBallot(ballot *ballotInfo) {
+	s.ballots[ballot.layer] = append(s.ballots[ballot.layer], ballot)
+	s.ballotRefs[ballot.id] = ballot
+}
+
+func (s *state) addBlock(block *blockInfo) {
+	layer := s.layer(block.layer)
+	layer.blocks = append(layer.blocks, block)
+	s.blockRefs[block.id] = block
+	s.updateRefHeight(layer, block)
+}
+
+func (s *state) updateRefHeight(layer *layerInfo, block *blockInfo) {
+	if layer.verifying.referenceHeight == 0 && layer.lid.After(s.evicted) {
+		layer.verifying.referenceHeight = s.layer(layer.lid.Sub(1)).verifying.referenceHeight
+	}
+	if block.height <= s.referenceHeight[block.layer.GetEpoch()] &&
+		block.height > layer.verifying.referenceHeight {
+		layer.verifying.referenceHeight = block.height
+	}
+}
+
+type (
+	blockVote struct {
+		*blockInfo
+		vote sign
+	}
+
+	baseInfo struct {
+		id    types.BallotID
+		layer types.LayerID
+	}
+
+	conditions struct {
+		baseGood bool
+		// is ballot consistent with local opinion within [base.layer, layer)
+		consistent bool
+		// set after comparing with local beacon
+		badBeacon bool
+		// any exceptions before base.layer
+		votesBeforeBase bool
+	}
+
+	ballotInfo struct {
+		id         types.BallotID
+		layer      types.LayerID
+		base       baseInfo
+		height     uint64
+		weight     weight
+		beacon     types.Beacon
```

Comment:
you are right, technically beacon matters only in ref ballot. but there are some nuances when beacon can be dropped from ref ballot. let me check that in the upcoming pr, as it will change when badBeacon is determined